### ISSUE TITLE
[no-inferrable-types] Don't flag `const` variables specifying non-literal types

### ISSUE THREAD

**Issue by ericbf** [typescript-eslint/typescript-eslint#2750]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-inferrable-types": ["error"]
  }
}
```

```TS
export const someVar: string = "value"
```

**Expected Result**

No error, as the specified type differs from what would be the inferred type.

**Actual Result**

"Type string trivially inferred from a string literal, remove type annotation."

Removing the type annotation results in a literal type of `"value"`, which will then throw errors if trying to compare to other string values, for variables that are replace at runtime or compile time.

**Additional Info**

I'd expect it not to flag cases where the specified type differs from the inferred type.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.31.0` |
| `@typescript-eslint/parser`        | `2.31.0` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.7.0` |

**Comment by bradzacher** [comment]
The whole point of the rule is to flag these cases.
That's literally all the rule does - flag cases where you've got an explicit type annotation on a literal value.

> which will then throw errors if trying to compare to other string values

Can you give examples?

**Comment by ericbf** [comment]
The rule flags (should flag, imo) cases where the specified type does not change the inferred type (e.g., when using a `let`).

Since typescript narrows the type to the literal type when using `const`, specifying a non-literal type changes the type of the variable, and should not be specified as "trivially inferable" as the inferred type is different.

```ts
export const someVar = "value"
export const otherVar: string = "value"

if (someVar === "not value") {
   ...
}

if (otherVar === "not value") {
    ...
}
```

[Playground for the example](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwgOwM71RAtsAagQyjgF44AiAN3wBsBXYMgKFElkRXTghgAtgoCUAFxx0UAJbIA5iXJU6DRo3EAzOAApMOQSWKkyybnHn0yASjgBvJQF8lqjdz4DCu-YfgmGF63D9xGGyA).

The above throws an error for `someVar` but not for `otherVar`, because `someVar` is the literal type `"value"`. If we have the `: string`, we basically tell TS to not use the literal type, but to keep the non-literal `string` type. It's a different behavior than not including the type specification.

Obviously in the example the const value won't change, but if we are trying to use a `const` while potentially changing the value at runtime (maybe during packaging), this will result in the literal type not being appropriate.

**Comment by bradzacher** [comment]
Typescript is doing the correct thing here. As far as it's concerned - the string is constant and that value.

If it's the case that you have some custom build-time/runtime systems which change the assumptions TS is making, then this is a required type annotation.

However as the rule isn't ever going to understand your non-standard system, this is where you can use a disable comment and explain why you're going against the convention and explain that it is a necessary type annotation.

Why is this much better? It clearly shows every developer on the codebase that this annotation exists for a good reason and should not be deleted.

**Comment by ericbf** [comment]
> Typescript is doing the correct thing here. As far as it's concerned - the string is constant and that value.

Don't know who said it wasn't "doing the correct thing"? The point was that by specifying a type different from the inferred type (`string` as opposed to the literal type), typescript understands that the comparison is between two strings instead of a string to a literal type.

Nextly, this isn't a discussion about packaging solutions or any supposed "non-standard" system or whatever. The point is that if the specified type is different from the inferred type, it's clearly not an "inferable type", as the inferred type differs from the specified type. The rule can understand that the type specification is different from the inferred type (in my understanding of the type system). If it is different, the rule could know that it's there for a reason. It could also potentially be under an option in the rule about if the specified type differs from the inferred type, it would not be flagged, or something along those lines.

So far as adding disabling comments, that's always an option when rules are not behaving as expected, but is usually not a good solution. It defeats the point of using the rule in the first place.

**Comment by bradzacher** [comment]
You're talking about this from a pure types perspective.
I'm talking about this from a practical perspective.

`const x: string = 'a'` widens the type from `'a'` to `string`. This is a correct statement.

From a pure types perspective - is this an inferrable type?
No - because `'a'` is not `string`.

From a **practical** perspective - is this classifiable as an inferrable type?
For 99.999% of the cases - yes.
I believe that there's actually only one case where manually widening the type will actually change something - when you're doing a logical comparison on strings.
For all other usages TS will automatically (and implicitly) widen the type to `string` at the usage site if it needs to.

> It defeats the point of using the rule in the first place.

**Strongly** disagree.
The entire point of disable comments is to either:
- workaround bugs so you're not blocked by linting
- clearly document known edge cases that a lint rule either cannot detect
- clearly document an intentional and purposeful violation of a lint rule

The first usage is the same reason things like `@ts-ignore` exist.
The second usage is something we should endeavour to reduce as much as possible.
The final usage is 100% completely valid and in no way defeats the point.

----

Look at it this way.
Imagine someone comes into your codebase and adds an explicit type annotation to all `const` variable declarations.
Would you look at this and say "they're explicit annotations, they must be there for a reason".
Or would you question _why_ every single `const` has an explicit annotation?

This rule exists to make you question - to make you think twice.
This is to help you reduce the amount of unnecessary type annotations in your codebase.

If you find there is a necessary type annotation - use a disable comment and explain its necessity.
Otherwise - delete it to make your codebase cleaner, and your typecheck faster.

**Comment by ericbf** [comment]
Alright—I'll accept that. Thanks for looking at it!
